lychee-lib/src/extract/markdown.rs

```diff
@@ -1,35 +1,131 @@
 
 /// Extract unparsed URL strings from a Markdown string.
-pub(crate) fn extract_markdown(input: &str) -> Vec<RawUri> {
+pub(crate) fn extract_markdown(input: &str, skip_code_blocks: bool) -> Vec<RawUri> {
+    // In some cases it is undesirable to extract links from within code blocks,
+    // which is why we keep track of entries and exits while traversing the input.
+    let mut inside_code_block = false;
+
     let parser = Parser::new(input);
     parser
-        .flat_map(|event| match event {
-            MDEvent::Start(Tag::Link(_, uri, _)) => {
-                vec![RawUri {
+        .filter_map(|event| match event {
+            // A link. The first field is the link type, the second the destination URL and the third is a title.
+            Event::Start(Tag::Link(_, uri, _)) => {
+                Some(vec![RawUri {
                     text: uri.to_string(),
                     // Emulate `<a href="...">` tag here to be compatible with
                     // HTML links. We might consider using the actual Markdown
                     // `LinkType` for better granularity in the future
                     element: Some("a".to_string()),
                     attribute: Some("href".to_string()),
-                }]
+                }])
             }
-            MDEvent::Start(Tag::Image(_, uri, _)) => {
-                vec![RawUri {
+            // An image. The first field is the link type, the second the destination URL and the third is a title.
+            Event::Start(Tag::Image(_, uri, _)) => {
+                Some(vec![RawUri {
                     text: uri.to_string(),
                     // Emulate `<img src="...">` tag here to be compatible with
                     // HTML links. We might consider using the actual Markdown
                     // `LinkType` for better granularity in the future
                     element: Some("img".to_string()),
                     attribute: Some("src".to_string()),
-                }]
+                }])
+            }
+            // A code block (inline or fenced).
+            Event::Start(Tag::CodeBlock(_)) => {
+                inside_code_block = true;
+                None
+            }
+            Event::End(Tag::CodeBlock(_)) => {
+                inside_code_block = false;
+                None
+            }
+
+            // A text node.
+            Event::Text(txt) => {
+                if inside_code_block && skip_code_blocks {
+                    None
+                } else {
+                    Some(extract_plaintext(&txt))
+                }
             }
-            MDEvent::Text(txt) => extract_plaintext(&txt),
-            MDEvent::Html(html) => extract_plaintext(&html.to_string()),
-            _ => vec![],
+
+            // An HTML node
+            Event::Html(html) => Some(extract_plaintext(&html.to_string())),
+
+            // An inline code node.
+            Event::Code(code) => {
+                println!("found code block: {code}");
```

Comment:
do we need to print this?

Reply:
Nope, that was some debug output that can be removed.

Reply:
Thanks for the heads-up.

Reply:
We'll probably have to tackle code blocks anyway, for now I removed the println.